operator/internal/controller/vectorized/cluster_controller_test.go

```diff
@@ -0,0 +1,386 @@
+	"github.com/redpanda-data/redpanda-operator/pkg/kube/kubetest"
+	"github.com/redpanda-data/redpanda-operator/pkg/otelutil/log"
+)
+
+func TestClusterControllerGolden(t *testing.T) {
+	tests := []struct {
+	name           string
+	clusterDef     func() *vectorizedv1alpha1.Cluster
+	reconcilerFunc func(t *testing.T, mgr manager.Manager) *ClusterReconciler
+	goldenFile     string
+	normalizerFunc func(ctx context.Context, testScheme *runtime.Scheme, cluster *vectorizedv1alpha1.Cluster, t *testing.T, ctl *kube.Ctl) []client.Object
+}{
+		{
+			name:           "minimal_cluster",
+			clusterDef:     minimalClusterDef,
+			reconcilerFunc: minimalReconciler,
+			normalizerFunc: minimalClusterNormalizer,
+			goldenFile:     "testdata/cluster-resources.golden.yaml",
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			// Unified scheme gives issues so we're using a home bake
+			testScheme := runtime.NewScheme()
+			for _, fn := range []func(s *runtime.Scheme) error{
+				clientgoscheme.AddToScheme,
+				certmanagerv1.AddToScheme,
+				vectorizedv1alpha1.AddToScheme,
+				apiextensionsv1.AddToScheme,
+			} {
+				require.NoError(t, fn(testScheme))
+			}
+			ctx := log.IntoContext(t.Context(), testr.New(t))
+
+			ctl := kubetest.NewEnv(t, kube.Options{
+				Options: client.Options{
+					Scheme: testScheme,
+				},
+			})
+
+			require.NoError(t, kube.ApplyAllAndWait(ctx, ctl, func(crd *apiextensionsv1.CustomResourceDefinition, err error) (bool, error) {
+				if err != nil {
+					return false, err
+				}
+
+				for _, cond := range crd.Status.Conditions {
+					if cond.Type == apiextensionsv1.Established {
+						return cond.Status == apiextensionsv1.ConditionTrue, nil
+					}
+				}
+
+				return false, nil
+			}, crds.All()...))
+
+			mgr, err := ctrl.NewManager(ctl.RestConfig(), manager.Options{
+				Logger:  testr.New(t),
+				Scheme:  testScheme,
+				Metrics: server.Options{BindAddress: "0"},
+			})
+			require.NoError(t, err)
+
+			reconciler := tt.reconcilerFunc(t, mgr)
+			err = ctrl.NewControllerManagedBy(mgr).
+				For(&vectorizedv1alpha1.Cluster{}).
+				Owns(&appsv1.StatefulSet{}).
+				Owns(&corev1.Service{}).
+				Named("cluster-" + fmt.Sprintf("%d", time.Now().UnixNano())).
+				WithOptions(controller.Options{
+					SkipNameValidation: ptr.To(true),
+					NeedLeaderElection: ptr.To(false),
+				}).
+				Complete(reconciler)
+			require.NoError(t, err)
+
+			mgrCtx, mgrCancel := context.WithCancel(ctx)
+			go func() {
+				require.NoError(t, mgr.Start(mgrCtx))
+			}()
+
+
+			cluster := tt.clusterDef()
+			require.NoError(t, ctl.Apply(ctx, cluster))
+			require.NoError(t, ctl.WaitFor(ctx, cluster, func(obj kube.Object, err error) (bool, error) {
+				if err != nil {
+					return false, err
+				}
+
+				cl := obj.(*vectorizedv1alpha1.Cluster)
+				fmt.Printf("conditions: %s", cl.Status.Conditions)
+				return len(cl.Status.Conditions) > 0, nil
+			}))
+
+			defer cleanupTest(t, ctx, ctl, cluster, mgrCancel)
```

Comment:
Prefer `t.Cleanup` to defer in tests. Also if you're tearing down the entire envtest each test run, there's not really any cleanup required as it cleans itself up by (at least in theory) shutting down the entire etcd and kube-apiserver process.